Set meta.mainProgram for executable components

chuahou opened the pull request:
This allows `nix run` to work correctly, so that it knows the correct name of the executable, as it would otherwise attempt to execute an executable named `project-name-exe-exeName` for the component `project-name:exe:exeName`, for example. It sets it to the empty string for non-executable components.

Fixes #1231.

---

Tests I ran include:
- [x] Tested by hand for exe, lib and test components on a skeleton project:
  - `meta.mainProgram` is set correctly, and
  - `nix run` works out of the box.
- [x] Full test suite (incl. impure) as described [in the docs](https://input-output-hk.github.io/haskell.nix/dev/tests.html), for `ghc902` (I ran `./tests.sh ghc902`).
  - Succeeds up until I encounter #1482.
- [ ] Currently running the full test suite for `ghc8107` due to the above issue.

---

As I'm completely new to haskell.nix, I'm not sure what else I should look out for or test. I based this off [@hamishmack's suggestion in #1231](https://github.com/input-output-hk/haskell.nix/issues/1231#issuecomment-1174734062), and if I'm not wrong the check component would inherit `meta` from `drv` and hence doesn't need `meta.mainProgram` to be set again.

michaelpj commented on builder/comp-builder.nix:
```diff
@@ -356,6 +356,7 @@ let
       description = package.synopsis or "";
       license = haskellLib.cabalToNixpkgsLicense package.license;
       platforms = if platforms == null then lib.platforms.all else platforms;
+      mainProgram = if haskellLib.isExecutableType componentId then exeName else "";
```
Can we just not include this attribute if it's not an executable? What does nix run do if it's the empty string as opposed to missing?

Worth a comment explaining what you said in the PR about nix run so we know what this is for in future.

michaelpj replied:
Is there something we can put there that would make it actually fail? What happens if we put `error "Component <component> is not an executable, you can't run it"`?

michaelpj replied:
I'm not sure what the idiomatic thing to do here is!

chuahou replied:
Setting it as
```
mainProgram = if haskellLib.isExecutableType componentId then exeName
              else throw "Component ${name} is not an executable, you can't run it.";
```
allows `nix run` to work on executables while showing that error message for non-executable components like `lib`s, but I'm not sure if this is a good idea since `[package].meta ? mainProgram` now returns `true` and the `[package].meta.mainProgram or [something else]` pattern no longer works. This may or may not be desirable, and I have no clue what the idiomatic thing to do here is either.

michaelpj replied:
Fair enough. It just seems a bit sad that the fallback case is "try and run an obviously non-existent executable" rather than "tell the user they can't do that".

chuahou replied:
Yeah. But I suppose that's true of any nix derivations independently from haskell.nix, and the iffiness comes from nixpkgs not having `meta.mainProgram` before but having it now, and wanting to support derivations that depend on old behaviour.

hamishmack commented on builder/comp-builder.nix:
```diff
@@ -356,6 +356,7 @@ let
       description = package.synopsis or "";
       license = haskellLib.cabalToNixpkgsLicense package.license;
       platforms = if platforms == null then lib.platforms.all else platforms;
+      mainProgram = if haskellLib.isExecutableType componentId then exeName else "";
```
Nice!  `exeName` includes the `extension` (`.exe` on windows for instance).  It does not look like [getExe](https://github.com/NixOS/nixpkgs/blob/8ef7523c8e11da7fd23e6b87371e7aa1eab718bb/lib/meta.nix#L141-L142) adds it so I guess including it is the right thing to do.

Can you please update so that it does not set `mainProgram` to `""`.  Something like:
```
  } // lib.optionalAttrs (haskellLib.isExecutableType componentId) {
    mainProgram = exeName;
  };
```

chuahou replied:
Got it, I'll change it so it doesn't set `mainProgram` for non-executable components. Thanks!

chuahou commented:
I've changed it in 323d3473bf22072cb0cea115d30f1b5779865627 to not set `mainProgram` for non-executable components and added a comment in the source as per the reviews above. I've verified it continues to work as expected with my skeleton project, and am waiting for `./tests.sh ghc8107` to run its course right now.
